src/eager/mod.rs

```diff
@@ -0,0 +1,835 @@
+                        name: CStr::from_ptr(c_name).to_str()?.to_string(),
+                        device_type: CStr::from_ptr(c_type).to_str()?.to_string(),
+                        memory_bytes: bytes,
+                        incarnation,
+                    });
+                }
+                Ok(devices)
+            })();
+            tf::TF_DeleteDeviceList(list);
+            result
+        }
+    }
+
+    /// Clears the internal caches in the context.
+    pub fn clear_cashes(&mut self) {
+        unsafe {
+            tf::TFE_ContextClearCaches(self.inner);
+        }
+    }
+}
+
+unsafe impl std::marker::Send for Context {}
+unsafe impl std::marker::Sync for Context {}
+
+/// A handle to a tensor on a device.
+///
+/// Constructing a TensorHandle requires a reference to an execute context so that the
+/// generated handle will not out live the context.
+/// ```
+/// # use tensorflow::Tensor;
+/// use tensorflow::eager::{ContextOptions, Context, ToHandle};
+///
+/// let opts = ContextOptions::new();
+/// let ctx = Context::new(opts).unwrap();
+///
+/// let t = Tensor::from(3i32);                // <- Tensor
+/// let h = t.to_handle(&ctx).unwrap();        // <- TensorHandle
+/// let v: Tensor<i32> = h.resolve().unwrap(); // <- Tensor
+/// assert_eq!(v[0], 3i32);
+/// ```
+///
+/// Since TensorHandle cannot be alive beyond the lifetime of the context, the following
+/// code will not compile.
+/// ```compile_fail
+/// # use tensorflow::Tensor;
+/// use tensorflow::eager::{ContextOptions, Context, ToHandle};
+///
+/// let h = {
+///     let opts = ContextOptions::new();
+///     let ctx = Context::new(opts).unwrap();
+///
+///     let t = Tensor::from(3);
+///     let h = t.to_handle(&ctx).unwrap();
+///     h
+/// };
+/// ```
+#[derive(Debug)]
+pub struct TensorHandle<'a> {
+    inner: *mut tf::TFE_TensorHandle,
+    // TensorHandle should not live longer than a given context.
+    ctx: PhantomData<&'a Context>,
+}
+impl<'a> Drop for TensorHandle<'a> {
+    fn drop(&mut self) {
+        unsafe {
+            tf::TFE_DeleteTensorHandle(self.inner);
+        }
+    }
+}
+
+impl<'a> TensorHandle<'a> {
+    /// Crate a TensorHandle from Tensor
+    pub fn new<T: TensorType>(_ctx: &'a Context, t: &Tensor<T>) -> Result<TensorHandle<'a>> {
+        let status = Status::new();
+        let inner = unsafe { tf::TFE_NewTensorHandle(t.inner().unwrap(), status.inner) };
+
+        if inner.is_null() {
+            Err(status)
+        } else {
+            Ok(TensorHandle {
+                inner,
+                ctx: PhantomData,
+            })
+        }
+    }
+
+    /// Returns the DataType that corresponds to this type.
+    pub fn data_type(&self) -> DataType {
+        unsafe { DataType::from_c(tf::TFE_TensorHandleDataType(self.inner)) }
+    }
+
+    /// Return a number of dimensions.
+    /// This function will block till the operation that produces the TensorHandle has completed.
+    pub fn num_dims(&self) -> Result<i32> {
```

Comment:
I think I can leave the return type as is, since the following code uses the same type (while it is referred to as `c_int`). This code is in graph.rs.


```rust
    /// Returns the number of dimensions of the Tensor referenced by `output`.
    ///
    /// If the number of dimensions in the shape is unknown, returns -1.
    ///
    /// Returns an error if:
    ///
    ///   * `output` is not in `graph`.
    pub fn num_dims<I: Into<Output>>(&self, output: I) -> Result<c_int> {
```